x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorConsumer.java

```diff
@@ -58,90 +63,120 @@
             @Override
             protected boolean removeEldestEntry(final Map.Entry<String, Boolean> eldest) {
                 return false == eldest.getKey().startsWith(todayISODate());
             }
-        }));
+        });
+        this.workQueue = new LinkedList<>();
     }
 
     @Override
-    public void accept(Collection<RoleDescriptor> effectiveRoleDescriptors) {
-        threadPool.generic().execute(() -> {
-            final SortedMap<String, AliasOrIndex> aliasOrIndexMap = clusterService.state().metaData().getAliasAndIndexLookup();
-            // executing the check asynchronously will not conserve the generated deprecation response headers (which is what we want,
-            // because it's not the request that uses deprecated features, but rather the role definition. Plus, due to caching, we can't
-            // reliably associate response headers to every request).
-            logDeprecatedPermission(effectiveRoleDescriptors, aliasOrIndexMap);
-        });
+    public synchronized void accept(Collection<RoleDescriptor> effectiveRoleDescriptors) {
+        final Collection<RoleDescriptor> uncachedEffectiveRoleDescriptors = new ArrayList<>();
+        for (RoleDescriptor roleDescriptor : effectiveRoleDescriptors) {
+            if (dailyRoleCache.add(buildCacheKey(roleDescriptor))) {
+                uncachedEffectiveRoleDescriptors.add(roleDescriptor);
+            }
+        }
+        if (false == uncachedEffectiveRoleDescriptors.isEmpty()) {
+            workQueue.add(uncachedEffectiveRoleDescriptors);
+            if (workQueue.size() == 1) {
+                threadPool.generic().execute(() -> {
+                    // executing the check asynchronously will not conserve the generated deprecation response headers (which is what we
+                    // want, because it's not the request that uses deprecated features, but rather the role definition. Furthermore, due to
+                    // caching, we can't reliably associate response headers to every request).
+                    logDeprecatedPermission(uncachedEffectiveRoleDescriptors);
+                    // this thread will continue to pop work items
+                    while (true) {
+                        // synchronize access for the next work item
+                        final Collection<RoleDescriptor> workItem = tryGetNextWorkItem();
+                        if (workItem == null) {
+                            break;
+                        }
+                        logDeprecatedPermission(workItem);
+                    }
+                });
+            }
+        }
     }
 
-    private void logDeprecatedPermission(Collection<RoleDescriptor> effectiveRoleDescriptors,
-            SortedMap<String, AliasOrIndex> aliasOrIndexMap) {
+    private synchronized Collection<RoleDescriptor> tryGetNextWorkItem() {
+        assert workQueue.poll() != null; // this was the processed item
+        return workQueue.peek();
```

Comment:
peek does not remove the item.

Reply:
If peek returns `null` there's no more work and the worker fades out.
Otherwise it returns the next work item, start working on it, but do not remove it - to signal the producer that a worker is humming - only remove it when the work is done.